feat: Network indexer (#2651)
nearprotocol/nearcore: chain/indexer/README.md

# NEAR Indexer

NEAR Indexer is a micro-framework, which provides you with a stream of blocks that are recorded on NEAR network.

NEAR Indexer is useful to handle real-time "events" on the chain.

NEAR Indexer is going to be used to build NEAR Explorer, augment NEAR Wallet, and provide overview of events in Rainbow Bridge.

See the [example](https://github.com/nearprotocol/nearcore/tree/master/tools/indexer/example) for further details.

## How to set up and test NEAR Indexer

Before you proceed, make sure you have the following software installed:
* [rustup](https://rustup.rs/) or Rust version that is mentioned in `rust-toolchain` file in the root of nearcore project.

### Localnet

Clone [nearcore](https://github.com/nearprotocol/nearcore)

```bash
$ git clone git@github.com:nearprotocol/nearcore.git
$ cd nearcore
$ cargo run --package neard --bin neard init
```

The above commands should initialize necessary configs and keys to run localnet

```bash
$ cd tools/indexer/example
$ cargo run --release
```

> After that you should see logs of every block produced by your localnet

Reviewer: ````suggestion
After the node is started, you should see logs of every block produced in your localnet. Get back to the code to implement any custom handling of the data flowing into the indexer.

Use [near-shell](https://github.com/near/near-shell) to submit transactions. For example, to create a new user you run the following command:

```
$ env NEAR_ENV=local near --keyPath ~/.near/localnet/validator_key.json create_account new-account.test.near --masterAccount test.near
```
````

# NEAR Indexer

NEAR Indexer is a micro-framework, which provides you with a stream of blocks that are recorded on NEAR network.

NEAR Indexer is useful to handle real-time "events" on the chain.

NEAR Indexer is going to be used to build NEAR Explorer, augment NEAR Wallet, and provide overview of events in Rainbow Bridge.

See the [example](https://github.com/nearprotocol/nearcore/tree/master/tools/indexer/example) for further details.

## How to set up and test NEAR Indexer

> Assuming you have all necessary tools installed (Rust, Cargo etc.)

Reviewer: @khorolets how does one know which tools are necessary if even we cannot state them explicitly? :smile: 

```suggestion
Before you proceed, make sure you have the following software installed:

* [rustup](https://rustup.rs/) or Rust version that is mentioned in `rust-toolchain` file in the root of nearcore project.
```

> ### Localnet

Reviewer: ```suggestion
### localnet
```

Clone [nearcore](https://github.com/nearprotocol/nearcore)

```bash
$ git clone git@github.com:nearprotocol/nearcore.git
$ cd nearcore
$ cargo run --package neard --bin neard init
```

The above commands should initialize necessary configs and keys to run localnet

```bash
$ cd tools/indexer/example
$ cargo run
```

After that you should see logs of every block produced by your localnet

> ### Betanet

Reviewer: ```suggestion
### betanet
```

To run the NEAR Indexer connected to Betanet we need to have configs and keys prepopulated, you can get them with the [nearup](https://github.com/near/nearup). Clone it and follow the instruction to run non-validating node (leaving account ID empty).

> Configs for beta net are in the `~/.near/betanet` folder. We need to ensure `"tracked_shards"` parameter in `~/.near/betaneta/config.json` set up properly.
> 
> It has to be

Reviewer: ```suggestion
Configs for betanet are in the `~/.near/betanet` folder. We need to ensure that NEAR Indexer follows all the necessary shards, so `"tracked_shards"` parameter in `~/.near/betaneta/config.json` needs to be configured properly. For example, with a single shared network, you just add the shard #0 to the list:
```

# NEAR Indexer

NEAR Indexer is a micro-framework, which provides you with a stream of blocks that are recorded on NEAR network.

NEAR Indexer is useful to handle real-time "events" on the chain.

NEAR Indexer is going to be used to build NEAR Explorer, augment NEAR Wallet, and provide overview of events in Rainbow Bridge.

See the [example](https://github.com/nearprotocol/nearcore/tree/master/tools/indexer/example) for further details.

## How to set up and test NEAR Indexer

Before you proceed, make sure you have the following software installed:
* [rustup](https://rustup.rs/) or Rust version that is mentioned in `rust-toolchain` file in the root of nearcore project.

### localnet

Clone [nearcore](https://github.com/nearprotocol/nearcore)

```bash
$ git clone git@github.com:nearprotocol/nearcore.git
$ cd nearcore

> $ cargo run --package neard --bin neard init

Reviewer: I suggest we add `--release` here to avoid double compilation given that we recommend compiling Indexer Example with `--release`.

```suggestion
$ env "NEAR_HOME=$HOME/.near/localnet" cargo run --release --package neard --bin neard init
```

```

> The above commands should initialize necessary configs and keys to run localnet

Reviewer: ```suggestion
The above commands should initialize necessary configs and keys to run localnet in `~/.near/localnet`.
```

```bash
$ cd tools/indexer/example

> $ cargo run --release /path/to/.near/config_dir/

Reviewer: ```suggestion
$ cargo run --release ~/.near/localnet/
```

```

After the node is started, you should see logs of every block produced in your localnet. Get back to the code to implement any custom handling of the data flowing into the indexer.

Use [near-shell](https://github.com/near/near-shell) to submit transactions. For example, to create a new user you run the following command:

```
$ env NEAR_ENV=local near --keyPath ~/.near/localnet/validator_key.json create_account new-account.test.near --masterAccount test.near
```

### betanet

To run the NEAR Indexer connected to betanet we need to have configs and keys prepopulated, you can get them with the [nearup](https://github.com/near/nearup). Clone it and follow the instruction to run non-validating node (leaving account ID empty).

> Configs for betanet are in the `~/.near/betanet` folder. We need to ensure that NEAR Indexer follows all the necessary shards, so `"tracked_shards"` parameter in `~/.near/betaneta/config.json` needs to be configured properly. For example, with a single shared network, you just add the shard #0 to the list:
> 
> ```
> ...
> "tracked_shards": [0],
> ...
> ```

Reviewer: ````suggestion
Configs for betanet are in the `~/.near/betanet` folder. We need to ensure that NEAR Indexer follows all the necessary shards and syncs all the blocks, so `"tracked_shards"` and `"block_fetch_horizon"` parameters in `~/.near/betanet/config.json` need to be configured properly. For example, with a single shared network, you just add the shard #0 to the list:

```
...
"consensus": {
  ...
  "block_fetch_horizon": 18446744073709551615,
  ...
},
...
"tracked_shards": [0],
...
```
````

After that we can run NEAR Indexer.

Follow to `nearcore` folder.

```bash
$ cd nearcore/tools/indexer/example

> $ cargo run --release

Reviewer: @bowenwang1996 Should the current master be able to play nicely with the current betanet?

@khorolets If that is not the case, one should patch beta branch with this PR to run it against betanet chain, and I would prefer to have a git command here in README applying the changes from this PR to whichever branch users want to.

Reviewer 2: > Should the current master be able to play nicely with the current betanet?

Yes

Author: So we are good to go without additional actions? If so, please resolve this conversation.

Reviewer: I mark it as resolved, but also want to ask you to confirm it locally

> $ cargo run --release

Reviewer: ```suggestion
$ cargo run --release ~/.near/betanet
```

# NEAR Indexer

NEAR Indexer is a micro-framework, which provides you with a stream of blocks that are recorded on NEAR network.

NEAR Indexer is useful to handle real-time "events" on the chain.

NEAR Indexer is going to be used to build NEAR Explorer, augment NEAR Wallet, and provide overview of events in Rainbow Bridge.

See the [example](https://github.com/nearprotocol/nearcore/tree/master/tools/indexer/example) for further details.

## How to set up and test NEAR Indexer

> Assuming you have all necessary tools installed (Python, Rust, Cargo etc.)

Reviewer: Let's be specific. Do we acually require Python?

### Localnet

Clone [nearcore](https://github.com/nearprotocol/nearcore)

```bash
$ git clone git@github.com:nearprotocol/nearcore.git
$ cd nearcore
$ cargo run --package neard --bin neard init
```

The above commands should initialize necessary configs and keys to run localnet

```bash
$ cd tools/indexer/example
$ cargo run
```

After that you should see logs of every block produced by your localnet

### Betanet

Clone [nearcore](https://github.com/nearprotocol/nearcore)

```bash
$ git clone git@github.com:nearprotocol/nearcore.git
$ cd nearcore
$ make release
```

This will take awhile to build `nearcore` (we are building `release` as it works a bit faster than debug)

To connect to Betanet we need to have a proper configs and keys.

Clone [nearup](https://github.com/near/nearup)

```bash
$ git clone git@github.com:near/nearup.git
$ cd nearup
$ ./nearup betanet --nodocker --binary-path path/to/nearcore/target/release
```

> After that you will be asked for your validator account ID, you can leave it empty as we are not going to validate.

Reviewer: Just reference nearup and say that one should follow the usual instructions running a non-validating node (leaving the account ID empty). This way we don't need to update our instructions every time the process changes.

BTW, the configs and keys are prepopulated automatically, which is the only reason we use nearup in the first place, so we may as well clearly state that the whole pre-running step is required to get the necessary configs prepared for us.

After some time you'll see a success message that says the node is running.

Now we need to stop the node.

```bash
$ ./nearup stop
```

Configs for beta net are in the `~/.near/betanet` folder. We need to ensure `"tracked_shards"` parameter in `~/.near/betaneta/config.json` set up properly.

It has to be

```
...
"tracked_shards": [0],
...
```

After that we can run NEAR Indexer.

Follow back to `nearcore` folder.

```bash
$ cd nearcore/tools/indexer/example

> $ cargo run

Reviewer: ```suggestion
$ cargo run --release
```

```

> After that you should see logs of every block produced by Betanet

Reviewer: ```suggestion
After the network is synced, you should see logs of every block produced in Betanet. Get back to the code to implement any custom handling of the data flowing into the indexer.
```
